rfcs/41-ui-navigation.md

```diff
@@ -0,0 +1,394 @@
+the position of our UI elements, but it also has knowledge of the logical
+organization of elements through the `Parent`/`Children` relationship. (in this
+case, children are specified in the parenthesis following the element name,
+look at `grid(...)`)
+
+Therefore, with the little information we provided through this hand wavy
+specification, we should already have everything setup for the navigation to
+just work™.
+
+
+### Dimensionality
+
+Let's go back to our ambition of using LT/RT for navigation now. 
+
+UI navigation is not just 2D, it's N-dimensional. The
+LT/RT to change tabs is like navigating in a 3rd orthogonal dimension to the
+dimensions you navigate with UP DOWN LEFT RIGHT inside the menus.
+
+I'll call those dimensions `Plane`s because it's easier to type than
+"Dimension", I'll limit the implementation to 3 planes. We can imagine truly
+exotic menu navigation systems with more than 3 dimensions, but I'll not
+worry about that. Our `Plane`s are:
+* `Plane::Menu`: Use LT/RT to move from one element to the next/previous
+* `Plane::Select`: Instead of emitting an `Action` event when left-clicking or
+  pressing A/B, go up-down that direction
+* `Plane::Physical`: Use the `Transform` positioning and navigate with Dpad or
+  arrow keys
+
+Each `NavCommand` moves the focus on a specific `Plane` as follow:
+* `Plane::Menu`: `Previous`, `Next`
+* `Plane::Select`: `Action`, `Cancel`
+* `Plane::Physical`: `MoveUp`, `MoveDown`, `MoveLeft`, `MoveRight`
+
+We should also be able to "loop" the menu, for example going LEFT of "soul"
+loops us back to "all" at the other side of the tab bar.
+
+### Specifying navigation dimensions
+
+We posited we could easily infer the physical layout and bake a navigation map
+automatically based on the `Transform` positions of the elements. This is not
+the case for the dimensions of our navigation. We can't magically infer the
+intent of the developer: They need to specify the dimensionality of their menus.
+
+Let's go back to the menu. In the example code, I refer to `:container` and
+`:navigable`. I've not explained yet what those are. Let's clear things up.
+
+A _navigable_ is an element that can be focused. A _container_ is a node entity
+that contains _navigables_ and 0 or 1 other _container_.
+
+![The previous tabbed menu example, but with _containers_ and _navigables_ highlighted](https://user-images.githubusercontent.com/26321040/140542768-4fdd5f23-2c2e-43c1-9fa4-cc11fe67c619.png)
+
+The _containers_ are represented as semi-transparent squares; the _navigables_
+are circles.
+
+In rust, it might look like this:
+```rust
+struct Container {
+  inner: Option<Box<Container>>,
+  siblings: NonEmptyVec<Navigable>,
+  active: SiblingIndex,
+  plane: Plane,
+}
+```
+Note: this is the data structure for the navigation algorithm, the ECS
+componenet called `Container` will probably look like this:
+```rust
+#[derive(Component)]
+struct Container {
+  plane: Plane,
```

Comment:
So, given that we're using separate entities for these containers already, and you could have multiple of them, I think the planes (or whatever bikeshed names) should be components of the `Plane<P>` type.

Then, you can have separate systems for each plane type (using a generic helper method on `App`, stealing tech from events) and navigate very quickly using query filters. This will also make the meaning of each container much clearer in various tools like `bevy_egui_inspector`, since we don't need to go digging into the values.

Then, since looping is defined on a per-plane basis, this should be defined in the helper method on `App`.

Reply:
I really like this suggestion! I'll keep in mind for the implementation.